# Задание 1
Загрузите датасет Wine из встроенных датасетов sklearn.datasets с помощью функции load_wine в
переменную data.

In [128]:
from sklearn.datasets import load_wine
import numpy as np
import pandas as pd

In [129]:
data = load_wine()

# Задание 2 
Полученный датасет не является датафреймом. Это структура данных, имеющая ключи
аналогично словарю. Просмотрите тип данных этой структуры данных и создайте список data_keys,
содержащий ее ключи.

In [111]:
type(data)

sklearn.utils.Bunch

In [110]:
data_keys = data.keys()

In [17]:
data_keys

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

# Задание 3
Просмотрите данные, описание и названия признаков в датасете. Описание нужно вывести в виде
привычного, аккуратно оформленного текста, без обозначений переноса строки, но с самими
переносами и т.д.

In [18]:
data['feature_names']

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

In [19]:
print(data['DESCR'])

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

# Задание 4 
Сколько классов содержит целевая переменная датасета? Выве
дите названия классов.

целевая переменная содержит 3 класса

In [20]:
data['target_names']

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

# Задание 5
На основе данных датасета (они содержатся в двумерном массиве Numpy) и названий признаков
создайте датафрейм под названием X.

In [21]:
X = pd.DataFrame(data['data'], columns = [data['feature_names']])

# Задание 6
Выясните размер датафрейма X и установите, имеются ли в нем пропущенные значения.

In [22]:
X.shape

(178, 13)

In [23]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 13 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   (alcohol,)                       178 non-null    float64
 1   (malic_acid,)                    178 non-null    float64
 2   (ash,)                           178 non-null    float64
 3   (alcalinity_of_ash,)             178 non-null    float64
 4   (magnesium,)                     178 non-null    float64
 5   (total_phenols,)                 178 non-null    float64
 6   (flavanoids,)                    178 non-null    float64
 7   (nonflavanoid_phenols,)          178 non-null    float64
 8   (proanthocyanins,)               178 non-null    float64
 9   (color_intensity,)               178 non-null    float64
 10  (hue,)                           178 non-null    float64
 11  (od280/od315_of_diluted_wines,)  178 non-null    float64
 12  (proline,)            

В данном датафрейме 13 столбцов, 178 строк. Пропущенных значений нету

# Задание 7
Добавьте в датафрейм поле с классами вин в виде чисел, имеющих тип данных numpy.int64.
Название поля - 'target'.

In [93]:
X['target'] = data['target'].astype(np.int64)

# Задание 8
Постройте матрицу корреляций для всех полей X. Дайте полученному датафрейму название
X_corr.

In [124]:
X_corr = X.corr()


In [125]:
X_corr


alcohol malic_acid       ash alcalinity_of_ash  \
alcohol                       1.000000   0.094397  0.211545         -0.310235   
malic_acid                    0.094397   1.000000  0.164045          0.288500   
ash                           0.211545   0.164045  1.000000          0.443367   
alcalinity_of_ash            -0.310235   0.288500  0.443367          1.000000   
magnesium                     0.270798  -0.054575  0.286587         -0.083333   
total_phenols                 0.289101  -0.335167  0.128980         -0.321113   
flavanoids                    0.236815  -0.411007  0.115077         -0.351370   
nonflavanoid_phenols         -0.155929   0.292977  0.186230          0.361922   
proanthocyanins               0.136698  -0.220746  0.009652         -0.197327   
color_intensity               0.546364   0.248985  0.258887          0.018732   
hue                          -0.071747  -0.561296 -0.074667         -0.273955   
od280/od315_of_diluted_wines  0.072343  -0.368710  0.003911         -0.276769   
proline                       0.643720  -0.192011  0.223626         -0.440597   
target                       -0.328222   0.437776 -0.049643          0.517859   

                             magnesium total_phenols flavanoids  \
alcohol                       0.270798      0.289101   0.236815   
malic_acid                   -0.054575     -0.335167  -0.411007   
ash                           0.286587      0.128980   0.115077   
alcalinity_of_ash            -0.083333     -0.321113  -0.351370   
magnesium                     1.000000      0.214401   0.195784   
total_phenols                 0.214401      1.000000   0.864564   
flavanoids                    0.195784      0.864564   1.000000   
nonflavanoid_phenols         -0.256294     -0.449935  -0.537900   
proanthocyanins               0.236441      0.612413   0.652692   
color_intensity               0.199950     -0.055136  -0.172379   
hue                           0.055398      0.433681   0.543479   
od280/od315_of_diluted_wines  0.066004      0.699949   0.787194   
proline                       0.393351      0.498115   0.494193   
target                       -0.209179     -0.719163  -0.847498   

                             nonflavanoid_phenols proanthocyanins  \
alcohol                                 -0.155929        0.136698   
malic_acid                               0.292977       -0.220746   
ash                                      0.186230        0.009652   
alcalinity_of_ash                        0.361922       -0.197327   
magnesium                               -0.256294        0.236441   
total_phenols                           -0.449935        0.612413   
flavanoids                              -0.537900        0.652692   
nonflavanoid_phenols                     1.000000       -0.365845   
proanthocyanins                         -0.365845        1.000000   
color_intensity                          0.139057       -0.025250   
hue                                     -0.262640        0.295544   
od280/od315_of_diluted_wines            -0.503270        0.519067   
proline                                 -0.311385        0.330417   
target                                   0.489109       -0.499130   

                             color_intensity       hue  \
alcohol                             0.546364 -0.071747   
malic_acid                          0.248985 -0.561296   
ash                                 0.258887 -0.074667   
alcalinity_of_ash                   0.018732 -0.273955   
magnesium                           0.199950  0.055398   
total_phenols                      -0.055136  0.433681   
flavanoids                         -0.172379  0.543479   
nonflavanoid_phenols                0.139057 -0.262640   
proanthocyanins                    -0.025250  0.295544   
color_intensity                     1.000000 -0.521813   
hue                                -0.521813  1.000000   
od280/od315_of_diluted_wines       -0.428815  0.565468   
proline            

# Задание 9
Создайте список high_corr из признаков, корреляция которых с полем target по абсолютному
значению превышает 0.5 (причем, само поле target не должно входить в этот список).

In [131]:
 high_corr = X_corr.loc[(np.abs(X_corr['target']) > 0.5) & (np.abs(X_corr['target']) < 1.0)]

ValueError: Cannot index with multidimensional key